src/models.jl

```diff
@@ -1,10 +1,78 @@
-function ising(g::AbstractGraph; h)
+_maybe_fill(x, n) = x
+_maybe_fill(x::Number, n) = fill(x, n)
+
+"""
+Random field J1-J2 Heisenberg model on a general graph
+"""
+function heisenberg(
+  g::AbstractGraph; J1=1.0, J2=0.0, h::Union{<:Real,Vector{<:Real}}=0
+)
+  h = _maybe_fill(h, nv(g))
   ℋ = OpSum()
-  for e in edges(g)
-    ℋ -= "Z", maybe_only(src(e)), "Z", maybe_only(dst(e))
+  if !iszero(J1)
+    for e in edges(g)
+      ℋ += J1 / 2, "S+", maybe_only(src(e)), "S-", maybe_only(dst(e))
+      ℋ += J1 / 2, "S-", maybe_only(src(e)), "S+", maybe_only(dst(e))
+      ℋ += J1, "Sz", maybe_only(src(e)), "Sz", maybe_only(dst(e))
+    end
   end
-  for v in vertices(g)
-    ℋ += h, "X", maybe_only(v)
+  if !iszero(J2)
+    # TODO, more clever way of looping over next to nearest neighbors?
```

Comment:
This would be nice functionality to have, and is worth splitting off into a separate function (`next_nearest_neighbors`?).

It looks like we could make use of [Graphs.neighborhood_dists](https://juliagraphs.org/Graphs.jl/dev/algorithms/connectivity/#Graphs.neighborhood_dists-Union{Tuple{T},%20Tuple{U},%20Tuple{AbstractGraph{T},%20Integer,%20Any},%20Tuple{AbstractGraph{T},%20Integer,%20Any,%20AbstractMatrix{U}}}%20where%20{U%3C:Real,%20T%3C:Integer}) to create a more general function for outputting the vertices that are the k-nearest neighbors of a specified vertex, for example:
```julia
function k_nearest_neighbors(g::AbstractGraph, v, d)
  nds = neighborhood_dists(g, v, d)
  # Filter `nds` for vertices with distances equal to `d`
end
```
Then `next_nearest_neighbors(g, v) = k_nearest_neighbors(g, v, 2)`.

Reply:
Sounds good. I assume it would make sense to add this in NamedGraphs.jl? I'll try to find some time to add a PR soon.

Reply:
Yeah, it would make sense to add to `NamedGraphs.jl`.

Reply:
Note that [Graphs.neighborhood_dists](https://juliagraphs.org/Graphs.jl/dev/algorithms/connectivity/#Graphs.neighborhood_dists-Union%7BTuple%7BT%7D,%20Tuple%7BU%7D,%20Tuple%7BAbstractGraph%7BT%7D,%20Integer,%20Any%7D,%20Tuple%7BAbstractGraph%7BT%7D,%20Integer,%20Any,%20AbstractMatrix%7BU%7D%7D%7D%20where%20%7BU%3C:Real,%20T%3C:Integer%7D) should now work as of [NamedGraphs](https://github.com/mtfishman/NamedGraphs.jl/pull/17).

Reply:
Tracking this issue in mtfishman/NamedGraphs.jl#22.